File: src/game_state.rs
use crate::barn::game::context::Context;
use crate::barn::game::state::State;
use crate::camera::Camera;
use crate::eye::Eye;
use crate::fire::Fire;
use crate::level_select_state::LevelSelectState;
use crate::physics::handle_collisions;
use crate::player::Player;
use crate::tile::Tile;
use sdl2::mixer::Chunk;
use sdl2::render::Texture;
use sdl2::render::TextureCreator;

use sdl2::keyboard::Keycode;
use sdl2::pixels::Color;
use sdl2::rect::Rect;
use sdl2::render::BlendMode;
use sdl2::render::WindowCanvas;

use std::cell::RefCell;
use std::collections::HashMap;
use std::fs;
use std::path::Path;
use std::rc::Rc;
use std::time::Instant;

pub const TILE_WIDTH: u32 = 64;
pub const TILE_HEIGHT: u32 = 64;

pub const TILE_CHARS: [char; 6] = ['*', '5', '5', '6', '8', '9'];

pub struct GameState/*<'a>*/ {
    pub level_path: String,
    pub won: bool,
    pub time: Instant,
    pub time_str: String,
    pub moves: u32,
    pub flames: Vec<Fire>,
    pub tiles: Vec<Tile>,
    pub blocks: Vec<Tile>,
    pub eyes: Vec<Eye>,
    pub player: Player,
    pub move_fx: Chunk,
    pub camera: Camera,
    // pub socket_tex: Texture<'a>,
}

impl State for GameState {
    fn update(&mut self, context: &mut Context) -> Option<Box<dyn State>> {
        if context.input.key_just_pressed(&Keycode::R) {
            self.load_level(String::from(&self.level_path), context);
        } else if context.input.key_just_pressed(&Keycode::Q) {
            return Some(Box::new(LevelSelectState {
                levels: HashMap::new(),
                options: Vec::new(),
                selected_option: 0,
                camera: Camera::new(),
                back_fx: sdl2::mixer::Chunk::from_file(Path::new("res/sound/back.ogg")).unwrap(),
                select_fx: sdl2::mixer::Chunk::from_file(Path::new("res/sound/select.ogg")).unwrap(),
                enter_fx: sdl2::mixer::Chunk::from_file(Path::new("res/sound/enter.ogg")).unwrap(),
                tiles: Vec::new(),
                eyes: Vec::new(),
            }));
        }

        for fire in self.flames.iter_mut() {
            fire.update();
        }

        // TODO: Refactor this.
        let new_tiles = &mut self.tiles.to_vec();
        for tile in self.tiles.iter_mut() {
            if tile.has_moved() {
                self.moves += 1;
            }
            tile.update(new_tiles, &self.move_fx);
        }

        // Check if the puzzle has been solved.
        self.won = true;
        for eye in self.eyes.iter_mut() {
            eye.update(&mut self.tiles);
            if !eye.solved {
                self.won = false;
            }
        }
        if !self.won {
            // Update the player.
            self.player.update(&mut context.input);
            handle_collisions(&mut self.player, &mut self.tiles, &self.move_fx);
            self.camera.focus(
                self.player.x + self.player.width as i32 / 2,
                self.player.y + self.player.height as i32 / 2,
            );
        } else {
            if self.time_str == "" {
                let elapsed_time = self.time.elapsed().as_secs();
                self.time_str = elapsed_time.to_string();
            }
            if context.input.key_just_pressed(&Keycode::Return) {
                return Some(Box::new(LevelSelectState {
                    levels: HashMap::new(),
                    options: Vec::new(),
                    selected_option: 0,
                    camera: Camera::new(),
                    back_fx: sdl2::mixer::Chunk::from_file(Path::new("res/sound/back.ogg")).unwrap(),
                    select_fx: sdl2::mixer::Chunk::from_file(Path::new("res/sound/select.ogg")).unwrap(),
                    enter_fx: sdl2::mixer::Chunk::from_file(Path::new("res/sound/enter.ogg")).unwrap(),
                    tiles: Vec::new(),
                    eyes: Vec::new(),
                }));
            }
        }

        // No state change has occured.
        None
    }

    fn draw(&mut self, context: &mut Context, canvas: &mut WindowCanvas) {
        self.camera.width = (canvas.output_size().unwrap().0) as i32;
        self.camera.height = (canvas.output_size().unwrap().1) as i32;
        
        canvas.set_draw_color(Color::RGB(0, 0, 0));
        canvas.clear();

        // Get all of the image assets.
        let tex_player = context.texture_manager.load("res/img/player.png").unwrap();
        let tex_shadow = context
            .texture_manager
            .load("res/img/drop_shadow.png")
            .unwrap();
        // Draw the scene.
        for tile in &mut self.tiles {
            if !tile.isblock
                && !tile.iswall
                && self
                    .camera
                    .is_object_visible(tile.x, tile.y, tile.width, tile.height)
            {
                tile.draw(
                    &context.texture_manager.load(&tile.texture).unwrap(),
                    &mut self.camera,
                    canvas,
                );
            }
        }
        self.player
            .draw_shadow(&tex_shadow, &mut self.camera, canvas);
        for tile in &mut self.tiles {
            if (tile.isblock || tile.iswall)
                && self
                    .camera
                    .is_object_visible(tile.x, tile.y, tile.width, tile.height)
            {
                tile.draw(
                    &context.texture_manager.load(&tile.texture).unwrap(),
                    &mut self.camera,
                    canvas,
                );
            }
        }
        for eye in self.eyes.iter_mut() {
            if self
                .camera
                .is_object_visible(eye.x, eye.y, eye.width, eye.height)
            {
                let tex_pupil = if eye.color == "blue" {
                    context
                        .texture_manager
                        .load("res/img/bluepupil.png")
                        .unwrap()
                } else {
                    if eye.color == "red" {
                        context
                            .texture_manager
                            .load("res/img/redpupil.png")
                            .unwrap()
                    } else {
                        context
                            .texture_manager
                            .load("res/img/greenpupil.png")
                            .unwrap()
                    }
                };
                // eye.draw(&mut self.socket_tex, &tex_pupil, &mut self.camera, canvas);
            }
        }
        self.player.draw(&tex_player, &mut self.camera, canvas);
        let tex_fire = context.texture_manager.load("res/img/fire2.png").unwrap();
        let tex_glow = context
            .texture_manager
            .load("res/img/fire_glow.png")
            .unwrap();
        for fire in self.flames.iter_mut() {
            fire.draw(&tex_fire, &tex_glow, &mut self.camera, canvas)
        }
        if self.won {
            canvas.set_draw_color(Color::RGBA(0, 0, 0, 150));
            canvas.set_blend_mode(BlendMode::Blend);
            canvas
                .fill_rect(Rect::new(
                    0,
                    0,
                    self.camera.width as u32,
                    self.camera.height as u32,
                ))
                .unwrap();
            let font = context.font_manager.load(&context.font_details).unwrap();
            let texture_creator = canvas.texture_creator();

            // Render the title.
            let title = font
                .render("Solved!")
                .blended(Color::RGBA(255, 255, 255, 255))
                .unwrap();
            let title_tex = texture_creator.create_texture_from_surface(&title).unwrap();
            canvas
                .copy(
                    &title_tex,
                    None,
                    Rect::new(
                        self.camera.width / 2 - 4 * title.size().0 as i32 / 2,
                        30,
                        title.size().0 * 4,
                        title.size().1 * 4,
                    ),
                )
                .unwrap();

            // Render time result.
            let time = font
                .render(&format!("Time: {} seconds", self.time_str))
                .blended(Color::RGBA(255, 255, 255, 255))
                .unwrap();
            let time_tex = texture_creator.create_texture_from_surface(&time).unwrap();
            canvas
                .copy(
                    &time_tex,
                    None,
                    Rect::new(
                        self.camera.width / 2 - 3 * time.size().0 as i32 / 2,
                        self.camera.height / 2 - 3 * time.size().1 as i32 / 2,
                        time.size().0 * 3,
                        time.size().1 * 3,
                    ),
                )
                .unwrap();

            // Render number of moves.
            let moves = font
                .render(&format!("Moves taken: {}", self.moves))
                .blended(Color::RGBA(255, 255, 255, 255))
                .unwrap();
            let moves_tex = texture_creator.create_texture_from_surface(&moves).unwrap();
            canvas
                .copy(
                    &moves_tex,
                    None,
                    Rect::new(
                        self.camera.width / 2 - 3 * moves.size().0 as i32 / 2,
                        self.camera.height / 2 + 4 * moves.size().1 as i32 / 2,
                        moves.size().0 * 3,
                        moves.size().1 * 3,
                    ),
                )
                .unwrap();

            // Render number of moves.
            let back = font
                .render("Press enter to go back")
                .blended(Color::RGBA(255, 255, 255, 255))
                .unwrap();
            let back_tex = texture_creator.create_texture_from_surface(&back).unwrap();
            canvas
                .copy(
                    &back_tex,
                    None,
                    Rect::new(
                        0,
                        self.camera.height - back.size().1 as i32 * 2,
                        back.size().0 * 2,
                        back.size().1 * 2,
                    ),
                )
                .unwrap();
        }
        canvas.present();
    }

    fn on_enter(&mut self, context: &mut Context) {
        self.load_level(String::from(&self.level_path), context);
        self.time = Instant::now();
    }

    fn on_exit(&mut self, context: &mut Context) {
        self.flames.clear();
    }

    fn get_name(&mut self) -> String {
        String::from("game")
    }
}

impl/*<'a>*/ GameState/*<'_>*/ {
    pub fn load_level(&mut self, level: String, context: &mut Context) {
        context.font_manager.load(&context.font_details).unwrap();
        let f = fs::read_to_string(level).expect("Could not load level!");
        let mut cury: i32 = 10;
        let mut temp_blocks: Vec<Tile> = Vec::new();
        let mut temp_eyes: Vec<Eye> = Vec::new();
        self.tiles.clear();
        self.eyes.clear();
        self.player = Player::new();
        let mut skip = true;
        for line in f.lines() {
            if skip {
                // Skip the first line of the file.
                skip = false;
                continue;
            }
            let mut curx: i32 = 10;
            for c in line.chars() {
                if TILE_CHARS.contains(&c) {
                    self.tiles.push(Tile {
                        texture: GameState::get_texture_name(c),
                        width: TILE_WIDTH,
                        height: TILE_HEIGHT,
                        x: curx,
                        y: cury,
                        targetx: curx,
                        targety: cury,
                        resistancex: 0,
                        resistancey: 0,
                        iswall: false,
                        isblock: false,
                    });
                    curx += TILE_WIDTH as i32;
                } else if c == 'f' {
                    let mut flame = Fire::new();
                    flame.x = curx + TILE_WIDTH as i32 / 2 - flame.width as i32 / 2;
                    flame.y = cury;
                    self.flames.push(flame);
                    self.tiles.push(Tile {
                        texture: String::from("res/img/torch.png"),
                        width: TILE_WIDTH,
                        height: TILE_HEIGHT,
                        x: curx,
                        y: cury,
                        targetx: curx,
                        targety: cury,
                        resistancex: 0,
                        resistancey: 0,
                        iswall: true,
                        isblock: false,
                    });
                    curx += TILE_WIDTH as i32;
                } else if c == 'x' {
                    self.tiles.push(Tile {
                        texture: String::from("res/img/grayblock.png"),
                        width: TILE_WIDTH,
                        height: TILE_HEIGHT,
                        x: curx,
                        y: cury,
                        targetx: curx,
                        targety: cury,
                        resistancex: 0,
                        resistancey: 0,
                        iswall: true,
                        isblock: false,
                    });
                    curx += TILE_WIDTH as i32;
                } else if c == 'b' {
                    temp_blocks.push(Tile {
                        texture: String::from("res/img/blueblock.png"),
                        width: TILE_WIDTH,
                        height: TILE_HEIGHT,
                        x: curx,
                        y: cury,
                        targetx: curx,
                        targety: cury,
                        resistancex: 30,
                        resistancey: 30,
                        iswall: false,
                        isblock: true,
                    });
                } else if c == 'g' {
                    temp_blocks.push(Tile {
                        texture: String::from("res/img/greenblock.png"),
                        width: TILE_WIDTH,
                        height: TILE_HEIGHT,
                        x: curx,
                        y: cury,
                        targetx: curx,
                        targety: cury,
                        resistancex: 30,
                        resistancey: 30,
                        iswall: false,
                        isblock: true,
                    });
                } else if c == 'r' {
                    temp_blocks.push(Tile {
                        texture: String::from("res/img/redblock.png"),
                        width: TILE_WIDTH,
                        height: TILE_HEIGHT,
                        x: curx,
                        y: cury,
                        targetx: curx,
                        targety: cury,
                        resistancex: 30,
                        resistancey: 30,
                        iswall: false,
                        isblock: true,
                    });
                } else if c == 'B' {
                    temp_eyes.push(Eye {
                        direction: String::from("left"),
                        x: curx,
                        y: cury,
                        width: TILE_WIDTH,
                        height: TILE_HEIGHT,
                        color: String::from("blue"),
                        solved: false,
                        deltax: 0,
                        deltay: 0,
                        anger: 0,
                    });
                } else if c == 'R' {
                    temp_eyes.push(Eye {
                        direction: String::from("left"),
                        x: curx,
                        y: cury,
                        width: TILE_WIDTH,
                        height: TILE_HEIGHT,
                        color: String::from("red"),
                        solved: false,
                        deltax: 0,
                        deltay: 0,
                        anger: 0,
                    });
                } else if c == 'G' {
                    temp_eyes.push(Eye {
                        direction: String::from("left"),
                        x: curx,
                        y: cury,
                        width: TILE_WIDTH,
                        height: TILE_HEIGHT,
                        color: String::from("green"),
                        solved: false,
                        deltax: 0,
                        deltay: 0,
                        anger: 0,
                    });
                } else if c == 'p' {
                    self.player.x = curx + (TILE_WIDTH / 2) as i32 - (self.player.width / 2) as i32;
                    self.player.y =
                        cury + 3 - (TILE_HEIGHT / 2) as i32 + (self.player.height / 2) as i32;
                } else if c == ' ' {
                    curx += TILE_WIDTH as i32;
                }
            }
            cury += TILE_HEIGHT as i32;
        }
        for tile in temp_blocks {
            self.tiles.push(tile);
        }
        for eye in temp_eyes {
            self.eyes.push(eye);
        }

        self.camera.minx = self.player.x + self.player.width as i32 / 2;
        self.camera.maxx = self.player.x + self.player.width as i32 / 2;
        self.camera.miny = self.player.y - self.player.height as i32 / 2;
        self.camera.maxy = self.player.y - self.player.height as i32 / 2;
    }

    fn get_texture_name(tile: char) -> String {
        if tile == '8' {
            return String::from("res/img/dbg_floor_shadow_top.png");
        } else if tile == '9' {
            return String::from("res/img/dbg_floor_shadow_ne_corner.png");
        } else if tile == '6' {
            return String::from("res/img/dbg_floor_shadow_right.png");
        } else if tile == '5' {
            return String::from("res/img/dbg_floor_shadow_corner.png");
        } else if tile == '*' {
            return String::from("res/img/dbg_floor.png");
        }
        String::from("res/img/dbg_floor.png")
    }
}

File: src/game_state.go
package mindseye

import (
	"bufio"
	"fmt"
	"os"
	"time"

	rl "github.com/gen2brain/raylib-go/raylib"
)

const TileWidth = 64
const TileHeight = 64

var TileChars = [6]string{"*", "5", "5", "6", "8", "9"}

type GameState struct {
	levelPath      string
	player         *Player
	camera         rl.Camera2D
	cameraMin      rl.Vector2
	cameraMax      rl.Vector2
	playerTexture  rl.Texture2D
	shadowTexture  rl.Texture2D
	completionFont rl.Font
	tileTextures   map[string]rl.Texture2D
	tiles          []Tile
	flames         []Fire
	eyes           []Eye
	won            bool
	moves          int
	startTime      time.Time
	solveTime      int64
}

func NewGameState(levelPath string) *GameState {
	return &GameState{levelPath: levelPath}
}

func (gs *GameState) Update(gc *GameContext, dt float32) (State, error) {
	if gs.player == nil {
		return nil, nil
	}
	if rl.IsKeyPressed(rl.KeyQ) {
		return NewLevelSelectState(), nil
	}
	if rl.IsKeyPressed(rl.KeyR) {
		if err := gs.loadLevel(gc, gs.levelPath); err != nil {
			return nil, err
		}
		return nil, nil
	}

	for i := range gs.flames {
		gs.flames[i].Update()
	}

	tileSnapshot := append([]Tile(nil), gs.tiles...)
	for i := range gs.tiles {
		if gs.tiles[i].HasMoved() {
			gs.moves++
		}
		gs.tiles[i].Update(tileSnapshot, dt)
	}

	for i := range gs.eyes {
		gs.eyes[i].Update(gs.tiles, dt)
	}

	gs.won = true
	for _, eye := range gs.eyes {
		if !eye.solved {
			gs.won = false
			break
		}
	}
	if !gs.won {
		gs.player.Update(dt)
		HandleCollisions(gs.player, gs.tiles, dt)
		gs.focusCamera()
	} else if gs.startTime.IsZero() {
		gs.startTime = time.Now()
	} else if gs.solveTime == 0 {
		gs.solveTime = int64(time.Since(gs.startTime).Seconds())
	} else if rl.IsKeyPressed(rl.KeyEnter) {
		return NewLevelSelectState(), nil
	}

	return nil, nil
}

func (gs *GameState) Draw(gc *GameContext) error {
	rl.BeginDrawing()
	defer rl.EndDrawing()
	rl.ClearBackground(rl.Black)
	rl.BeginMode2D(gs.camera)

	if gs.tileTextures == nil {
		gs.tileTextures = make(map[string]rl.Texture2D)
	}
	for _, tile := range gs.tiles {
		if tile.isWall || tile.isBlock {
			continue
		}
		gs.drawTile(tile)
	}
	for _, tile := range gs.tiles {
		if !tile.isWall && !tile.isBlock {
			continue
		}
		gs.drawTile(tile)
	}

	for _, eye := range gs.eyes {
		var pupil string
		switch eye.color {
		case "red":
			pupil = "res/img/redpupil.png"
		case "green":
			pupil = "res/img/greenpupil.png"
		default:
			pupil = "res/img/bluepupil.png"
		}
		socketTexture := gs.getLevelTexture("res/img/socket.png")
		pupilTexture := gs.getLevelTexture(pupil)
		eye.Draw(socketTexture, pupilTexture)
	}

	if gs.player != nil {
		gs.player.DrawShadow(gs.shadowTexture)
		gs.player.Draw(gs.playerTexture)
	}

	flameTexture := gs.getLevelTexture("res/img/fire2.png")
	glowTexture := gs.getLevelTexture("res/img/fire_glow.png")
	for i := range gs.flames {
		gs.flames[i].Draw(flameTexture, glowTexture)
	}

	rl.EndMode2D()
	if gs.won {
		gs.drawCompletionOverlay()
	}

	return nil
}

func (gs *GameState) drawCompletionOverlay() {
	width := int32(rl.GetScreenWidth())
	height := int32(rl.GetScreenHeight())
	rl.DrawRectangle(0, 0, width, height, rl.NewColor(0, 0, 0, 150))

	title := "Solved!"
	titleSize := int32(48)
	titlePosition := centeredTextPosition(gs.completionFont, title, float32(titleSize), width, height/6)
	rl.DrawTextEx(gs.completionFont, title, titlePosition, float32(titleSize), 0, rl.White)

	timeText := fmt.Sprintf("Time: %d seconds", gs.solveTime)
	movesText := fmt.Sprintf("Moves taken: %d", gs.moves)
	bodySize := float32(28)
	rl.DrawTextEx(gs.completionFont, timeText, centeredTextPosition(gs.completionFont, timeText, bodySize, width, height/2), bodySize, 0, rl.White)
	rl.DrawTextEx(gs.completionFont, movesText, centeredTextPosition(gs.completionFont, movesText, bodySize, width, height*3/5), bodySize, 0, rl.White)

	prompt := "Press enter to go back"
	promptSize := float32(22)
	rl.DrawTextEx(gs.completionFont, prompt, centeredTextPosition(gs.completionFont, prompt, promptSize, width, height*9/10), promptSize, 0, rl.White)
}

func centeredTextPosition(font rl.Font, text string, size float32, screenWidth, y int32) rl.Vector2 {
	measure := rl.MeasureTextEx(font, text, size, 0)
	return rl.NewVector2(float32(screenWidth)/2-measure.X/2, float32(y))
}

func (gs *GameState) OnEnter(gc *GameContext) error {
	gs.player = NewPlayer()
	gs.player.pos = rl.NewVector2(400, 300)
	gs.camera = rl.NewCamera2D(
		rl.NewVector2(400, 300),
		rl.NewVector2(0, 0),
		0,
		1,
	)
	gs.cameraMin = rl.NewVector2(0, 0)
	gs.cameraMax = rl.NewVector2(0, 0)

	gs.playerTexture = rl.LoadTexture("res/img/player.png")
	rl.SetTextureFilter(gs.playerTexture, rl.FilterPoint)

	gs.shadowTexture = rl.LoadTexture("res/img/drop_shadow.png")
	rl.SetTextureFilter(gs.shadowTexture, rl.FilterPoint)
	gs.completionFont = rl.LoadFont("res/fonts/VeniceClassic.ttf")
	rl.SetTextureFilter(gs.completionFont.Texture, rl.FilterPoint)

	levelPath := gs.levelPath
	if levelPath == "" {
		levelPath = "res/levels/level1.txt"
	}
	gs.levelPath = levelPath
	return gs.loadLevel(gc, levelPath)
}

func (gs *GameState) OnExit(gc *GameContext) error {
	if gs.playerTexture.ID != 0 {
		rl.UnloadTexture(gs.playerTexture)
		gs.playerTexture = rl.Texture2D{}
	}

	if gs.shadowTexture.ID != 0 {
		rl.UnloadTexture(gs.shadowTexture)
		gs.shadowTexture = rl.Texture2D{}
	}
	if gs.completionFont.Texture.ID != 0 {
		rl.UnloadFont(gs.completionFont)
		gs.completionFont = rl.Font{}
	}
	for path, texture := range gs.tileTextures {
		if texture.ID != 0 {
			rl.UnloadTexture(texture)
		}
		delete(gs.tileTextures, path)
	}

	return nil
}

func (gs *GameState) getLevelTexture(path string) rl.Texture2D {
	texture, ok := gs.tileTextures[path]
	if !ok {
		texture = rl.LoadTexture(path)
		rl.SetTextureFilter(texture, rl.FilterPoint)
		gs.tileTextures[path] = texture
	}
	return texture
}

func (gs *GameState) drawTile(tile Tile) {
	texture := gs.getLevelTexture(tile.texture)
	tile.Draw(texture)
}

func (gs *GameState) drawLevelTexture(path string, pos rl.Vector2) {
	gs.drawAnimatedTexture(
		path,
		rl.NewRectangle(0, 0, float32(TileWidth), float32(TileHeight)),
		pos,
		rl.NewVector2(TileWidth, TileHeight),
	)
}

func (gs *GameState) drawAnimatedTexture(path string, source rl.Rectangle, pos, size rl.Vector2) {
	texture, ok := gs.tileTextures[path]
	if !ok {
		texture = rl.LoadTexture(path)
		rl.SetTextureFilter(texture, rl.FilterPoint)
		gs.tileTextures[path] = texture
	}
	destination := rl.NewRectangle(
		pos.X,
		pos.Y,
		size.X,
		size.Y,
	)
	rl.DrawTexturePro(texture, source, destination, rl.NewVector2(0, 0), 0, rl.White)
}

func (gs *GameState) loadLevel(gc *GameContext, levelName string) error {
	file, err := os.Open(levelName)
	if err != nil {
		return err
	}
	defer file.Close()

	gs.tiles = nil
	gs.flames = nil
	gs.eyes = nil
	gs.won = false
	gs.moves = 0
	gs.startTime = time.Now()
	gs.solveTime = 0
	gs.player = NewPlayer()

	scanner := bufio.NewScanner(file)
	skip := true
	const levelOrigin = int32(10)
	cury := levelOrigin
	levelMaxX := float32(0)
	for scanner.Scan() {
		if skip {
			skip = false
			continue
		}
		line := scanner.Text()
		curx := levelOrigin
		for _, char := range line {
			switch char {
			case '*', '5', '6', '8', '9':
				gs.tiles = append(gs.tiles, Tile{
					texture:    tileTextureName(char),
					bb:         rl.NewRectangle(float32(curx), float32(cury), TileWidth, TileHeight),
					targetPos:  rl.NewVector2(float32(curx), float32(cury)),
					resistance: 30,
				})
				curx += TileWidth
			case 'f':
				fire := NewFire()
				fire.pos = rl.NewVector2(
					float32(curx+TileWidth/2)-fire.width/2,
					float32(cury),
				)
				gs.flames = append(gs.flames, fire)
				gs.tiles = append(gs.tiles, Tile{
					texture:    "res/img/torch.png",
					bb:         rl.NewRectangle(float32(curx), float32(cury), TileWidth, TileHeight),
					targetPos:  rl.NewVector2(float32(curx), float32(cury)),
					resistance: 30,
					isWall:     true,
				})
				curx += TileWidth
			case 'x':
				gs.tiles = append(gs.tiles, Tile{
					texture:    "res/img/grayblock.png",
					bb:         rl.NewRectangle(float32(curx), float32(cury), TileWidth, TileHeight),
					targetPos:  rl.NewVector2(float32(curx), float32(cury)),
					resistance: 30,
					isWall:     true,
				})
				curx += TileWidth
			case 'b', 'g', 'r':
				texture := map[rune]string{
					'b': "res/img/blueblock.png",
					'g': "res/img/greenblock.png",
					'r': "res/img/redblock.png",
				}[char]
				gs.tiles = append(gs.tiles, Tile{
					texture:    "res/img/dbg_floor.png",
					bb:         rl.NewRectangle(float32(curx), float32(cury), TileWidth, TileHeight),
					targetPos:  rl.NewVector2(float32(curx), float32(cury)),
					resistance: 30,
				})
				gs.tiles = append(gs.tiles, Tile{
					texture:    texture,
					bb:         rl.NewRectangle(float32(curx), float32(cury), TileWidth, TileHeight),
					targetPos:  rl.NewVector2(float32(curx), float32(cury)),
					resistance: 30,
					isBlock:    true,
				})
			case 'B', 'R', 'G':
				gs.tiles = append(gs.tiles, Tile{
					texture:    "res/img/dbg_floor.png",
					bb:         rl.NewRectangle(float32(curx), float32(cury), TileWidth, TileHeight),
					targetPos:  rl.NewVector2(float32(curx), float32(cury)),
					resistance: 30,
				})
				color := map[rune]string{'B': "blue", 'R': "red", 'G': "green"}[char]
				gs.eyes = append(gs.eyes, Eye{
					pos:       rl.NewVector2(float32(curx), float32(cury)),
					color:     color,
					direction: "left",
				})
			case 'p':
				gs.player.pos = rl.NewVector2(
					float32(curx+TileWidth/2-gs.player.width/2),
					float32(cury+3-TileHeight/2+gs.player.height/2),
				)
			case ' ':
				curx += TileWidth
			}
		}
		if float32(curx) > levelMaxX {
			levelMaxX = float32(curx)
		}
		cury += TileHeight
	}

	if err := scanner.Err(); err != nil {
		return err
	}
	gs.cameraMin = rl.NewVector2(float32(levelOrigin), float32(levelOrigin))
	gs.cameraMax = rl.NewVector2(levelMaxX, float32(cury))
	gs.focusCamera()
	return nil
}

func (gs *GameState) focusCamera() {
	if gs.player == nil {
		return
	}

	focus := rl.NewVector2(
		gs.player.pos.X+float32(gs.player.width)/2,
		gs.player.pos.Y+float32(gs.player.height)/2,
	)
	focusCamera(&gs.camera, focus, gs.cameraMin, gs.cameraMax)
}

func tileTextureName(tile rune) string {
	switch tile {
	case '8':
		return "res/img/dbg_floor_shadow_top.png"
	case '9':
		return "res/img/dbg_floor_shadow_ne_corner.png"
	case '6':
		return "res/img/dbg_floor_shadow_right.png"
	case '5':
		return "res/img/dbg_floor_shadow_corner.png"
	default:
		return "res/img/dbg_floor.png"
	}
}
